### ISSUE TITLE
Handling of partially satisfied prerequisites

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc-flow#4256]
In Cylc 7 partially satisfied prerequisites would eventually (by runahead limiting) stall a suite and prevent the scheduler from shutting down. (Although, so did fully unsatisfied prerequisites! - due to pre-spawning waiting tasks before they are needed)

In Cylc 8 we now have proper "spawn on demand" in that task proxies only appear in the n=0 task pool once *all* of their task-prerequisites are satisfied. Until then, for tasks with more than one prerequisite, prerequisite satisfaction is tracked invisibly (the fact that they are stored in task proxy objects is really just an implementation detail).

If there is nothing else to run, according to the graph, the scheduler shuts down. However, it logs a warning if any partially satisfied prerequisites are present.

## question

(Raised in the last meeting by @dpmatthews)

Should we consider it an error if the scheduler has nothing else to run but there are partially satisfied prerequisites present, and stay up as stalled rather than shut down?

**Comment by hjoliver** [comment]
## answer?

> Should we consider it an error if the scheduler has nothing else to run but there are partially satisfied prerequisites present, and stay up as stalled rather than shut down?

In my opinion, no.

While it is quite possible for a user to make a mistake by triggering a task that only leads on to a sub-graph when they intended the whole graph to run, it is also perfectly legit to do that intentionally, so it's not possible for Cylc to to determine which is which and we shouldn't automatically assume it was an error.

**Comment by hjoliver** [comment]
[Cut-and-paste from my defense of the status quo on Element chat!]

> Example:
```
[scheduling]                                      
  [[graph]]                                       
    R1 = "foo => bar & baz => qux"                
[runtime]                                         
   [[foo, bar, baz, qux]]                         
       script = sleep 10  
```
> If I `cylc play` this as normal, or do `cylc play --start-task=foo.1`, all four tasks will run then the scheduler will shut down.
But if I do `cylc play --start-task=bar.1`, only `bar.1` will run then the scheduler will shut down.
> 
> But this will be logged at shutdown:
```
2021-06-11T12:04:31+12:00 WARNING - Partially satisfied prerequisites left over:
        qux.1 is waiting on:
        * baz.1 succeeded
2021-06-11T12:04:32+12:00 INFO - DONE
```
> IMO that's the correct thing to do. qux.1 should appear in n=0 (active task) only when/if it becomes ready to run, which it never does because baz.1 has no reason to be spawned in this scenario.
baz and qux will be visible in the n=1 window of course.
>
>In Cylc 7 this would stall the suite largely for pragmatic reasons, although people may have come to think that's what should happen. Pragmatically, we had pre-spawned waiting tasks proxies at baz.1 and qux.1 and we couldn't distinguish between those that were "supposed to run" and those that according to the graph should not have been spawned in the first place (because their upstream parents did not run or whatever). So we stalled and left users to decide what to do.
>
> Of course we could decide to prevent shut down if there are partially satisfied prerequisites in Cylc 8. But I think that would be annoying if you want to trigger a task and have the scheduler proceed to do exactly what the graph says it should do, for some sub-graph.
> 
> So I think we should just educate users in this difference between Cylc 7 and Cylc 8 (and explain why the new way actually fixes a bug in Cylc 7!):
> - they should understand the structure of their graph - obviously!
>  - if they trigger a single task, or a few tasks, instead of starting at cycle point, they should see what the graph says will happen as a result
>  - if they don't want that to happen, they should trigger some other task instead. In the example above, start the flow from foo.1 instead of bar.1; or start from both bar.1 and baz.1.
>
> Note if you trigger bar.1 then decide you really did want qux.1 (and presumably baz.1) to run as well, there's several ways to achieve that. E.g.:
```
$ cylc play --pause d1/run1  # restart and pause
$ cylc trigger --reflow d1/run1 baz.1
$ cylc play d1/run1  # release
```
> (This will result in baz.1 and then qux.1 running, because partially satisfied prerequisites persist across restart)
>
> However, we should possibly make partially satisfied prerequisites more visible to users somehow. Most of the time they do not indicate anything is wrong (they will become satisfied as the workflow runs). Ultimately they should be attached to n>0 tasks in the UI where appropriate. But we could also consider regular checks (in the scheduler) and info or warnings in the UI for unsatisfied prerequisites that persist for "too long".

**Comment by dpmatthews** [comment]
> > Should we consider it an error if the scheduler has nothing else to run but there are partially satisfied prerequisites present, and stay up as stalled rather than shut down?

In my opinion, yes.

> While it is quite possible for a user to make a mistake by triggering a task that only leads on to a sub-graph when they intended the whole graph to run, it is also perfectly legit to do that intentionally, so it's not possible for Cylc to to determine which is which and we shouldn't automatically assume it was an error.

I agree it's not necessarily an error. However, that doesn't mean it makes sense for Cylc to just shutdown in the same way it would if the workflow had completed normally. If the user had done this intentionally it's not a big deal for the workflow to stall in this case. However, if it wasn't intentional the stall is an essential indication of a problem - relying on the user noticing a warning message in the log file is not sufficient.

Consider this workflow:
```
[scheduling]
    [[graph]]
        R1 = """
             foo => bar => qux
             foo:fail => baz => qux
             """
[runtime]
    [[foo, bar, baz, qux]]
        script = sleep 10
```
Currently this will just run foo and bar and then shutdown. The user intended qux to run but has made a mistake in the graph. It's too easy for a user to miss this (especially if this is part of a complex workflow or if it's something that only happens in some failure scenario).

Consider this workflow:
```
[scheduling]
    cycling mode = integer
    initial cycle point = 1
    [[graph]]
        P1 = """
             bar[-P1] => foo => bar & baz
             baz[-P1] => baz
             """
[runtime]
    [[foo,bar,baz]]
        script = sleep 2
```
If I run this with `cylc play --starttask=bar.2` then baz will never run and the workflow will currently shutdown (not stall) after `bar.7` runs due to the runahead limit (if #4258 gets merged I think it would run forever with no indication of a problem). Again, I think it's clear this isn't what the user would want and this would be a very easy mistake to make within a complex workflow.

In my opinion we have to treat any tasks with partially satisfied dependencies as an error in the graph and we should, therefore, stall if there are no more tasks to run. Otherwise I think there is a significant danger that graphing errors (either errors in the actual graph or mistakes made with triggering/re-flow) will go unnoticed by users (something which could never happen with Cylc 7).

**Comment by oliver-sanders** [comment]
Brain dump :)

#### My thoughts:

* Shutting down with a log message is not enough.
  * Few users ever look in the log.
  * Those who do only do so when investigating issues.
* The Cylc 7 "stalled" behaviour is useful.
  * Stalled used to approximately mean "workflow cannot continue without user intervention".
  * In these situations leaving the workflow running makes good sense.
  * We can communicate this nicely in the UI (e.g. badge on the workflow icon, with a warning icon in GScan, with a status indicator in the workflow toolbar).
  * These warnings are valuable for development and production alike.
  * In Cylc 8 workflows only "stall" if there are unhandled task failures left in the pool.
* [Whether rightly or wrongly] radical change in graph execution would likely become a reputational issue for us.
  * We can't expect users to understand the implementation details of SoD or SoS.
  * The implications of the change from SoD to SoS are too complex to explain to the majority of users.
  * We have a hell of a lot of Cylc 7 suites and we don't have enough resource to evaluate them to flush out SoD side-effects.
  * "It don't do what it used to" will become a problem for us.
* Graph branching is great, we don't want to sacrifice that flexibility.

So on the whole I find myself leaning in Dave's direction as maintaining this Cylc 7 functionality is more critical for us than introducing new Cylc 8 functionalities.

#### Generalisation of the issue:

I think Dave's assumption was that between unhandled failed tasks and partially satisfied prerequisites we have the required indicators to detect runtime issues.

However, here's an example of a runtime issue which wouldn't be flagged by either indicator:

```ini
[scheduling]
    [[graph]]
        R1 = a:x => b
[runtime]
    [[a]]
        script = true
        [[[outputs]]]
            x = x
```

Cylc 7: Task "a" runs, the workflow stalls with task "b" in the graph.

Cylc 8: Task "a" runs, the workflow hits its final cycle point and shuts down (`--no-detach` exit code 0). The workflow is not restartable, there is no warning that "b" did not run.

Cylc 7 provided this assurance:

> **Cylc 7:** All scheduled tasks are expected to be run unless explicitly removed by `cylc remove` or with a suicide trigger. If a scheduled task is not run the graph execution will halt, either due to the suite stalling or the graph execution hitting the runahead limit.

With the new graph branching functionality of Cylc 8 this assurance has been replaced by:

> **Cylc 8:** Some or all of the scheduled tasks will be run, the scheduler will shut down when the "workflow front" runs out.

On the whole this difference isn't an issue, but sometimes it is.

I would summarise these discrepancies as:

> Cylc 8 is liable to interpret execution and graph configuration errors as branch graphing.

(or at least the graph flexibility that branching has given us has resulted in the situation where graphs can terminate early because Cylc can't tell what the user intended to be run in this situation).

I.E Hillary's statement:

> it's not possible for Cylc to to determine which is which and we shouldn't automatically assume it was an error.

#### The crux of the problem:

I think the crux of the issue is that graph branching is implicit rather than explicit which means that:

* Graph branching is being applied to Cylc 7 graphs which might not be ready for it.
* The scheduler is unable to tell what the intended behaviour was.
* The scheduler cannot necessarily tell a stalled workflow from a completed workflow.

#### Quick catch-all solution?

If the graph branching was explicit it would become an opt-in feature. This means we could potentially maintain the assurances of Cylc 7 whilst providing the flexibility of Cylc 8.

Take the following implicit graph branching example:

```
a => b
a:fail => recover
b | recover => c
```

The intention here is to define two different execution pathways for which we would expect at least one to be invoked:

```
a
case (a)
    :succeed
        => b
    :fail
        => recover
b | recover => c
```

If we rewrite this graph to make the branching explicit by using a new trigger symbol, `?=>`, to mark a "branching dependency" like so:

```
a ?=> b
a:fail ?=> recover
b | recover => c
```

Then Cylc would be able to infer the following:

* "c" *must* be run because it is triggered by a "regular" dependency.
* "b" *might* be run because it is triggered by an "branching" dependency.
* "recover" *might* be run because it is triggered by an "branching" dependency.

Cylc wouldn't be able to infer that "b" and "recover" belong to different branches or understand that the output of task "a" is the switch (because the trigger could be a complex conditional) so Cylc wouldn't be able to enforce that one branch *or* the other was run, however, that's not so critical.

This could potentially be simple to implement, I think we would record the dependency type (regular/optional) in a task's "post requisites". When the task changes state we would go through the "post-requisites", if the requisite is a "regular" dependency and the state change has blocked the pathway to this requisite then we have a stall condition.

#### Idealised branching solution

As originally proposed in https://github.com/cylc/cylc-flow/issues/3226 graph branching would be accompanied by subgraph notation to provide Cylc with full awareness of the tasks on either branch:

```
a {
    :succeed
        a1 => a2
    :fail
        b1 => b2 => b3
        recover<b>
} => c
```

I.E:

```
# success branch
a:succeeded => a1 => a2 => c
# failure branch
a:fail => b1 => b2 => b3
a:fail => recover<b>
b3 & recover<b> => c
```

**Comment by hjoliver** [comment]
I love a good meaty discussion like this!

> Shutting down with a log message is not enough.
> - Few users ever look in the log.
> - Those who do only do so when investigating issues.

Agreed, and I did say above *"we could make partially satisfied prerequisites more visible somehow"* including *"...warnings in the UI"*.

However, it seems a bit perverse to me to keep a scheduler process alive and spinning its wheels indefinitely if there is nothing more for it to do, when we could easily make our tools (namely, `cylc scan` and the UI gscan panel) show the relevant information about stopped workflows instead, i.e. *I shut down with unsatisfied prerequisites* - and flag that as prominently as you like.

In fact I'd argue that's safer. If the user is away for a few days (or whatever) and the host goes down, it makes no difference if we are showing all the pertinent information about stopped workflows rather than relying on users finding stalled ones still running.

> The workflow is not restartable, 

Not true, we can already do that like this:
- `cylc play --pause <wf>`
- `cylc trigger --reflow <wf> bar.10`
- `cylc release <wf>`

And we could easily make that a one-step process of course.

**Comment by hjoliver** [comment]
> I think Dave's assumption was that between unhandled failed tasks and partially satisfied prerequisites we have the required indicators to detect runtime issues.
> 
> However, here's an example of a runtime issue which wouldn't be flagged by either indicator:

Agreed, and your example is exactly the kind of thing I was referring to above with this statement:

> Note also, the exact same problem (assuming it is unintentional) can occur in cases where only single prerequisites are involved, in which case no partially satisfied prerequisites will be present to signal what happened. (c.f. wholly unsatisfied waiting tasks causing a stall in Cylc 7).

But these cases do not necessarily imply an error either:
```ini
[scheduling]
    [[graph]]
        R1 = a:x => b
[runtime]
    [[a]]
        script = """
           if [[ new-data-available ]]; then
               # run b to process data
               cylc message x
           fi  # else we're done, no need to run b
        """
        [[[outputs]]]
            x = x
```

As you note, there's no way to detect this as a stall under SoD. So stalling with partially-satisfied prerequisites would amount to inconsistent handing of this kind of situation, depending on the number of prerequisites involved (i.e. one, or more than one).

**Comment by hjoliver** [comment]
> So on the whole I find myself leaning in Dave's direction as maintaining this Cylc 7 functionality is more critical for us than introducing new Cylc 8 functionalities

Well that's always a balancing act. I tend to come down on the other side of the fence because I think the damage is greater in the long term from sticking with sub-optimal behaviour to avoid a little short term hassle. This cruft can mount up over time! In my experience,  users are willing to change for well-justified reasons, even if they don't fully understand tor care about the reasons.

But in any case I don't think what we're discussing here falls far into that category at all. It doesn't strike me as at all "radical" to suggest that the flow should follow what the graph says, even if the graph is "wrong" with respect to the user's actual unknowable intentions - users should expect the same from anything they do on a computer.  If that occasionally causes a workflow to shut down prematurely (with respect to the aforementioned unknowable intentions) when it would have stalled in Cylc 7, well a) the Cylc 8 logic is much easy to justify and we can we make it easy to restart if necessary as discussed above; and b) by the same token there there are  other scenarios where Cylc 7 would have stalled wrongly and Cylc 8 "just works" (e.g. Cylc 7 stalls downstream of failed tasks).

**Comment by hjoliver** [comment]
> Cylc 7 provided this assurance:
> 
>    Cylc 7: All scheduled tasks are expected to be run unless explicitly removed by cylc remove or with a suicide trigger. If a scheduled task is not run the graph execution will halt, either due to the suite stalling or the graph execution hitting the runahead limit.

That rather oversells Cylc 7 - because "all scheduled tasks" means "all spawned tasks" which includes some that should not have been spawned (hence the need for suicide triggers) **and** it excludes some that should have been spawned but weren't (because of the implicit dependence on previous-instance submit).

So that "assurance" is really a two-sided coin, even if you think the Cylc 7 stall behaviour is desirable in the cases were it happens to be helpful (albeit for the wrong reasons!)

**Comment by hjoliver** [comment]
Now, the extended graph syntax ideas:

> Idealised branching (subgraph notation)

I've liked this idea ever since it was first raised, but I think it's fair to say we can't get that done by 8.0 (and it is somewhat less of an advance now that we have spawn-on-demand and don't need to worry about suiciding out unused sub-graphs).  Not that I think you were suggesting this for 8.0. Plus, not sure this qualifies as avoiding radical changes in graph execution :grin: 

> Quick catch-all solution? (`?=>` triggers)

This is an interesting idea :+1:, a way to expose something of what the user actually intends, to Cylc.  It would even work for single-prerequisite cases:
```
# foo is expected to generate output x, and bar is expected to run:
foo:x => bar  

# foo might generate output x, and bar should run only if it does:
foo:x ?=> bar
```

But if we introduce this, all triggers in existing Cylc 7 suites become "expected", so any branching will cause a "stall"?

(Stall in quotes now, because whether the scheduler stays up in a "stalled" state, or shuts down but exposes the final unsatisfied expected prerequisites to `cylc scan` is really a separate question, with or without these new triggers).

**Comment by dwsutherland** [comment]
My 2 cents:

> If the user had done this intentionally it's not a big deal for the workflow to stall in this case.

Perhaps some point in the graph is designed to stall, as a way to require human decision. (not just in the extreme case of a military application!)
Maybe an external trigger would be more appropriate, but some might like to just interact with the graph/UI.

As an ops person, we may not want to change the graph in the running workflow if the source dir is in flux (since you wouldn't be able to restart/play a stalled workflow without changing it)...
We also, often put in changes where the workflow will stall as a new task is missing a previously run (and despawned) task output.

**Comment by hjoliver** [comment]
> Perhaps some point in the graph is designed to stall, as a way to require human decision.

That's not a bad idea in its own right, perhaps.  But I don't think deliberately making a graphing error is the right way to do it!

**Comment by oliver-sanders** [comment]
> The workflow is not restartable,

>> Not true, we can already do that like this:

In my example the workflow would have hit the final cycle point so would not be restartable.

> not sure this qualifies as avoiding radical changes in graph execution 😁

We would have to use special syntax for subgraphs, I suggested `{ }`, in the original issue Ben suggested `( )`. Subgraphs would be a new opt-in feature so wouldn't change the execution of existing graphs.

Yes, was suggesting for Cylc 9, we don't have enough time for subgraphs right now, to implement them properly (rather than just expanding out the subgraphs in the config) would mean having nested a graph execution (necessary for fancy things like spinning off subsuites required to get real value out of this feature). Would need to be able to handle branching within the subgraphs, etc, big work for another day.

> But if we introduce this, all triggers in existing Cylc 7 suites become "expected", so any branching will cause a "stall"?

Yes, branching in Cylc 7 suites without suicide triggers would continue cause a stall so the old behaviour would be maintained.

If you want to allow the fancy new Cylc 8 branching you upgrade to the new trigger type and remove the redundant suicide triggers.

This really just puts the branching mechanism on rails so that branching can only occur in places where the user intended it to. This allows Cylc to tell the difference between an intended branch and a graph/execution error.

**Comment by hjoliver** [comment]
> 
> >  The workflow is not restartable,
> >
> > >       Not true, we can already do that like this:
> > >
> In my example the workflow would have hit the final cycle point so would not be restartable.

Yes it is, I actually tried it with your example, just to make sure :grin:

**Comment by hjoliver** [comment]
> The issue with branching ATM is that it changes the interpretation of Cylc 7 graphs and thus their execution.

Meh, not really. What exactly is different in terms of graph execution? In one case (Cylc 7) the workflow can stall (with unsatisfied waiting tasks present) and the scheduler stays up. In the other (Cylc 8) it runs out of tasks (possibly with partially satisfied prerequisites present) and the scheduler shuts down. But in both cases the same parts of the graph actually execute. And in both cases we could decide to keep the scheduler up or shut it down when it stalls or runs out of tasks to run. And in both cases, we could have `cylc scan` present the relevant shutdown status info and allow users to restart and trigger parts of the graph that didn't execute the first time because their prerequisites didn't get satisfied.

The only major difference in graph execution is Cylc 7 can stall unnecessarily, e.g. downstream of failed tasks, or if the user fails  to add suicide triggers for branching. Hopefully no one is relying on that sort of a stall.  Plus Cylc 8 can't (without your new `?=>` triggers) present information about unsatisfied single prerequisites.

**Comment by hjoliver** [comment]
We should seriously think about implementing `?=>` triggers though.

Taking all existing Cylc 7 triggers as "must" ones could be tricky because suicide triggers don't actually do anything in Cylc 8 (the task proxies they're supposed to remove do not get spawned, of course). We're essentially just retaining them to make existing graphs validate OK. I guess we should be able to repurpose them to turn "must" triggers to "might" though.

**Comment by oliver-sanders** [comment]
> Yes it is, I actually tried it with your example, just to make sure 😁

au contraire, it is restartable in the way that any flow that has hit the FCP is "restartable".

When you restart the flow it starts up then immediately shuts down before you get a chance to do anything to it.

```console
$ cylc play bar -n
...
2021-06-24T13:27:45+01:00 INFO - [a.1] -released from runahead
...
2021-06-24T13:27:48+01:00 INFO - [a.1] -Succeeded with outputs not completed:
	  x
2021-06-24T13:27:48+01:00 INFO - Workflow shutting down - AUTOMATIC
2021-06-24T13:27:48+01:00 INFO - DONE
$ cylc play bar -n
2021-06-24T13:27:49+01:00 INFO - Run: (re)start=1 log=1
2021-06-24T13:27:49+01:00 INFO - Cylc version: 8.0b2.dev
...
2021-06-24T13:27:49+01:00 INFO - Workflow shutting down - AUTOMATIC
2021-06-24T13:27:50+01:00 INFO - DONE
```

Technically you can `cylc play --pause` the workflow then `cylc trigger`, however, the triggered task will not run (because the flow is paused) and when you try to resyme it the flow will shut down before submitting the job.

(discussion for another thread)

**Comment by oliver-sanders** [comment]
> suicide triggers don't actually do anything in Cylc 8 

Good point, I think it can be handled.

Suicide triggers infer branching so we should be able to safely infer that any suicide triggered tasks are branched tasks:

1. Make a list of suicide triggered tasks.
2. Loop over dependencies to find those that trigger suicideable tasks.
3. Change these to `?=>` triggers in memory.

So basically do this on load in `cylc.config`:

```diff
 a => b
-a:fail => recover
+a:fail ?=> recover
 a => !recover
 b | recover => c
```

**Comment by dpmatthews** [comment]
Too much here for me to respond to at the moment but I remain convinced the Cylc 7 stall behaviour is essential.

> We should seriously think about implementing ?=> triggers though.

Agreed.

> When the task changes state we would go through the "post-requisites", if the requisite is a "regular" dependency and the state change has blocked the pathway to this requisite then we have a stall condition.

I'm assuming this would replace the special treatment of unhandled failed tasks. The task remains in the n=0 window if any of it's post-requisites are blocked. This would then eventually cause a stall. Is that right?

> Suicide triggers infer branching so we should be able to safely infer that any suicide triggered tasks are branched tasks

I need to give this some further thought ...

**Comment by hjoliver** [comment]
> au contraire, it is restartable in the way that any flow that has hit the FCP is "restartable"... you can cylc play --pause the workflow then cylc trigger, however, the triggered task will not run (because the flow is paused) and when you try to resyme it the flow will shut down before submitting the job.

ne pas au contraire! (?) 

Using your `a:x => b` example above, and following the procedure I gave above, which is exactly what you describe (restart with `--pause`, then trigger, then resume) the triggered task runs normally on resume.

```
2021-06-25T07:56:52+12:00 INFO - Paused on start up
2021-06-25T07:56:52+12:00 INFO - PAUSING the workflow now
2021-06-25T07:56:52+12:00 INFO - Scheduler: url=tcp://niwa-1007823l:43097/ pid=31718
2021-06-25T07:56:52+12:00 INFO - Workflow publisher: url=tcp://niwa-1007823l:43041
2021-06-25T07:56:52+12:00 INFO - Run: (re)start=1 log=1
2021-06-25T07:56:52+12:00 INFO - Cylc version: 8.0b2.dev
2021-06-25T07:56:52+12:00 INFO - Run mode: live
2021-06-25T07:56:52+12:00 INFO - Initial point: 1
2021-06-25T07:56:52+12:00 INFO - Final point: 1
2021-06-25T07:57:04+12:00 INFO - Processing 1 queued command(s)
2021-06-25T07:57:04+12:00 INFO - (no parent) spawned b.1 K
2021-06-25T07:57:04+12:00 INFO - Command succeeded: force_trigger_tasks(['b.1'], reflow=False)
2021-06-25T07:57:13+12:00 INFO - Processing 1 queued command(s)
2021-06-25T07:57:13+12:00 INFO - RESUMING the workflow now
2021-06-25T07:57:13+12:00 INFO - Command succeeded: resume()
2021-06-25T07:57:13+12:00 INFO - [b.1] -released from runahead
2021-06-25T07:57:13+12:00 INFO - Queue released: b.1
2021-06-25T07:57:13+12:00 INFO - [b.1] -submit-num=01, host=niwa-1007823l
2021-06-25T07:57:13+12:00 INFO - [b.1] -triggered off []
2021-06-25T07:57:14+12:00 INFO - [b.1] status=preparing: (internal)submitted at 2021-06-25T07:57:14+12:00  for job(01) flow(K)
2021-06-25T07:57:14+12:00 INFO - [b.1] -health check settings: submission timeout=None, polling intervals=PT15M,...
2021-06-25T07:57:15+12:00 INFO - [b.1] status=submitted: (received)started at 2021-06-25T07:57:14+12:00  for job(01) flow(K)
2021-06-25T07:57:15+12:00 INFO - [b.1] -health check settings: execution timeout=None, polling intervals=PT15M,...
2021-06-25T07:57:16+12:00 INFO - [b.1] status=running: (received)succeeded at 2021-06-25T07:57:15+12:00  for job(01) flow(K)
2021-06-25T07:57:16+12:00 INFO - Workflow shutting down - AUTOMATIC
2021-06-25T07:57:16+12:00 INFO - DONE
```

**Comment by hjoliver** [comment]
Summary from offline chat and meeting:

The proposed new notation probably solves most of our problems:
- `a:x ?=> b` is pure spawn-on-demand: if `a:x` does not get generated, `b` does not get spawned
- `a:x => b` is a "required dependency": when `a` finishes `b` will be spawned whether or not `a:x` was generated; if `a:x` was not generated, `b` will be present in the scheduler task pool with an unsatisfied prerequisite - similar to the partially satisfied multiple prerequisite case. 
- works for conditionals too: if `a:x | b:y => c` then `c` will be spawned when either of `a` or `b` finish, with or without the satisfied prerequisite (which can still potentially be satisfied later if the other upstream out gets generated)
 
@dpmatthews to think this through and report back on:
- any other problems or nasty edge cases?
- will we still need suicide triggers in certain situations?
- can this simplify our unhandled-failed task handling?
  - for `foo => bar`: if `foo` fails we'll be able to identify the unsatisfied `bar` as a problem even if we don't keep the failed `foo` in the pool ... but still convenient to keep it in?
  - should we implement something like `foo:fail => NULL` to indicate that failure doesn't matter for graph-leaf tasks?

**Comment by dpmatthews** [comment]
Way forward proposed in cylc/cylc-admin#128

**Comment by oliver-sanders** [comment]
Should be closed by #4343
